src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -822,31 +824,50 @@ DatabaseShardImp::import(Database& source)
                     shard->getBackend(),
                     nullptr,
                     nullptr,
-                    lastStoredLedger))
+                    recentStored))
                 {
                     break;
                 }
 
                 if (!shard->store(ledger))
                     break;
+
+                if (seq == lastLedgerSeq(shardIndex))
+                    lastLedgerHash = ledger->info().hash;
+
                 if (shard->isBackendComplete())
                 {
-                    JLOG(j_.debug()) <<
-                        "shard " << shardIndex << " was successfully imported";
+                    // Store shard final key
```

Comment:
If we get here we fully expect that `lastLedgerHash` should have been set, right?

Suppose a bug occurred where that was not the case.  Seems like we'd prefer to catch the bug rather than storing a bad shard final key?  Consider checking `lastLedgerHash == beast::zero` (and logging and breaking if true).  Or you could put `lastLedgerHash` in an `optional` if you want to be really clean.

If you follow this path you may also need to do something with line 873, since it also checks `shard->isBackendComplete()` and probably expects the shard final key to be successfully stored.

Just a thought, not a demand.

Reply:
@scottschurr I've addressed the issue, thanks!